Switch to @hyperjump/json-schema-coverage for schema test coverage

jdesrosiers opened the pull request:
Since our last update to the JSON Schema test coverage script, I've been working on a more complete solution and it's finally ready to share! I was able to generate istanbul coverage documents with the information collected from the @hyperjump/json-schema EvaluationPlugin we introduced in the last update. That allows us to generate familiar looking reports. I also figured out how to integrate it with Vitest, so it's super easy to use. I even included support for YAML out-of-the-box. See, [@hyperjump/json-schema-coverage](https://github.com/hyperjump-io/json-schema-coverage) for more details and examples.

**WARNING**: Because this does a more thorough job than the current solution, it will no longer report 100% coverage. More tests will need to be added. The biggest difference is that I added branch coverage support. Instead of just checking that the keyword was covered, it now also tracks whether both the true and false result of a keyword are covered.

**WARNING**: In the current version, we added a hack to ignore `#/$defs/schema` in `schema.yaml` because it isn't possible to cover that location if we're testing against `schema-base.yaml`. This new package doesn't have a solution for ignoring things. We could add a test that uses `schema.yaml` to cover that location, or look into some way to ignore coverage in the new solution.

**NOTE**: I think the `dev` branch is the right place to make this PR, but please let me know if that's wrong. The PR template says `main`, but the coverage tests don't exist on that branch.

---

- [ ] schema changes are included in this pull request
- [ ] schema changes are needed for this pull request but not done yet
- [x] no schema changes are needed for this pull request

ralfhandl commented on vitest.config.mjs:
```diff
@@ -2,7 +2,12 @@ import { defineConfig } from 'vitest/config'
 
 export default defineConfig({
   test: {
+    coverage: {
```
Setting [coverage thresholds](https://vitest.dev/config/#coverage-thresholds) to 100 seems to have no effect.

We want the tests to fail if schema coverage falls below 100% to avoid adding schema features without test cases.

jdesrosiers replied:
I got threshold enforcement through vitest working, so I removed nyc.

ralfhandl reviewed (commented):
@jdesrosiers Really nice!

Unfortunately we'd need some additions to use it, see comments.

jdesrosiers commented on tests/schema/schema.test.mjs:
```diff
@@ -60,10 +40,9 @@ describe("v3.1", () => {
           const instance = parseYamlFromFile(`${fixtures}/fail/${entry.name}`);
-          const output = validateOpenApi(instance, BASIC);
-          expect(output.valid).to.equal(false);
+          await expect(instance).to.not.matchJsonSchema("https://spec.openapis.org/oas/3.1/schema-base/WORK-IN-PROGRESS");
```
> Will these negative test cases count into the coverage?

Yes

> They should not.

Why not? The branch coverage feature ensures that your tests cover both the failing and passing cases of keywords. For example, branch coverage makes sure that for `type: object` you have a test that asserts that the instance passes when it's an object and another test that asserts that the instance fails when the document is something else, like `42`. You can only do that with a negative assertion.

If the test fails (either the positive tests or the negative tests) it will still record the coverage, but Vitest won't generate reports. Reports will only be generated if all the tests pass.

jdesrosiers commented:
It looks like defining a vocabulary isn't quite working, so this isn't working on the v3.2-dev branch yet.

jdesrosiers commented:
I fixed the problem with registering the vocabulary, so this now works for all branches including v3.2.

I also rebased to fix the conflict.

As far as I know, the only thing missing at this point is adding additional tests to cover the parts the new solution catches that the old one didn't. I'll leave that part to someone else.

ralfhandl commented on package.json:
```diff
@@ -15,7 +15,7 @@
     "build": "bash ./scripts/md2html/build.sh",
     "build-src": "npm run validate-markdown && bash ./scripts/md2html/build.sh src && bash ./scripts/schema-publish.sh src",
-    "test": "c8 --100 vitest --watch=false && bash scripts/schema-test-coverage.sh",
+    "test": "c8 --100 vitest run --coverage && nyc check-coverage",
```
```suggestion
    "test": "c8 --100 vitest run --coverage && ([ \"$BASE\" = \"dev\" ] || nyc check-coverage)",
```
We need to skip `nyc` for pull requests targeting the `dev` branch

ralfhandl commented:
> I don't have a windows machine to test locally, so let me know if there are any more issues.

Works like a charm, and you even got rid of the dependency to `nyc` 😄

ralfhandl commented on vitest.config.mjs:
```diff
@@ -1,8 +1,17 @@
+    globalSetup: ["tests/schema/oas-schema.mjs"],
+    coverage: {
+      include: ["src/schemas/validation/**/*.yaml"],
+      thresholds: process.env.BASE !== "dev" ? {
```
This is elegant!
